### PR TITLE
FIP-60 - Fuse guardian

### PR DESCRIPTION
**Pull request by Joeysantoro**
Creates a Fuse pause/borrow cap admin wrapper which gives access to Guardian and TribalChiefAdmins the ability to pause pools and set caps.

It also includes a utility for pausing/capping by underlying rather than the cToken. This is important because the cToken address cannot be pre-computed so this feature allows atomic deployment/pausing/capping.

### CODE REVIEW

File: contracts/feirari/FuseGuardian.sol [hunk 1]
```diff
@@ -0,0 +1,109 @@
...
+        comptroller._setMarketSupplyCaps(cTokens, newSupplyCaps);
+    }
+
+    function _setMarketSupplyCapsByUnderlying(address[] calldata underlyings, uint[] calldata newSupplyCaps) external {
+        _setMarketSupplyCaps(_underlyingToCTokens(underlyings), newSupplyCaps);
+    }
+
+    function _underlyingToCTokens(address[] memory underlyings) internal view returns (CToken[] memory) {
+        CToken[] memory cTokens = new CToken[](underlyings.length);
+        for (uint256 i = 0; i < underlyings.length; i++) {
```
Comment (transmissions11) [inline contracts/feirari/FuseGuardian.sol:42]: unchecked

File: contracts/feirari/FuseGuardian.sol [hunk 2]
```diff
@@ -0,0 +1,109 @@
...
+            address cToken = comptroller.cTokensByUnderlying(underlyings[i]);
+            require(cToken != address(0), "cToken doesn't exist");
+            cTokens[i] = CToken(cToken);
+        }
+        return cTokens;
```
Comment (transmissions11) [inline contracts/feirari/FuseGuardian.sol:47]: if u used a named return arg u could just assign cTokens to it

File: contracts/feirari/FuseGuardian.sol [hunk 3]
```diff
@@ -0,0 +1,109 @@
...
+      */
+    function _setMarketSupplyCaps(CToken[] memory cTokens, uint[] memory newSupplyCaps) public isGovernorOrGuardianOrAdmin {
+        comptroller._setMarketSupplyCaps(cTokens, newSupplyCaps);
+    }
+
+    function _setMarketSupplyCapsByUnderlying(address[] calldata underlyings, uint[] calldata newSupplyCaps) external {
+        _setMarketSupplyCaps(_underlyingToCTokens(underlyings), newSupplyCaps);
+    }
+
+    function _underlyingToCTokens(address[] memory underlyings) internal view returns (CToken[] memory) {
```
Comment (transmissions11) [inline contracts/feirari/FuseGuardian.sol:40]: this is only used by functions that take calldata so could make calldata

File: contracts/feirari/FuseGuardian.sol [hunk 4]
```diff
@@ -0,0 +1,109 @@
...
+    }
+
+    // ************ BORROW GUARDIAN FUNCTIONS ************
+    /**
+      * @notice Set the given supply caps for the given cToken markets. Supplying that brings total underlying supply to or above supply cap will revert.
+      * @dev Admin or borrowCapGuardian function to set the supply caps. A supply cap of 0 corresponds to unlimited supplying.
+      * @param cTokens The addresses of the markets (tokens) to change the supply caps for
+      * @param newSupplyCaps The new supply cap values in underlying to be set. A value of 0 corresponds to unlimited supplying.
+      */
+    function _setMarketSupplyCaps(CToken[] memory cTokens, uint[] memory newSupplyCaps) public isGovernorOrGuardianOrAdmin {
```
Comment (transmissions11) [inline contracts/feirari/FuseGuardian.sol:32]: this is only used by functions that take calldata so could make calldata

File: contracts/feirari/FuseGuardian.sol [hunk 5]
```diff
@@ -0,0 +1,109 @@
...
+        return cTokens;
+    }
+
+    /**
+      * @notice Set the given borrow caps for the given cToken markets. Borrowing that brings total borrows to or above borrow cap will revert.
+      * @dev Admin or borrowCapGuardian function to set the borrow caps. A borrow cap of 0 corresponds to unlimited borrowing.
+      * @param cTokens The addresses of the markets (tokens) to change the borrow caps for
+      * @param newBorrowCaps The new borrow cap values in underlying to be set. A value of 0 corresponds to unlimited borrowing.
+      */
+    function _setMarketBorrowCaps(CToken[] memory cTokens, uint[] memory newBorrowCaps) public isGovernorOrGuardianOrAdmin {
```
Comment (transmissions11) [inline contracts/feirari/FuseGuardian.sol:56]: see above

File: contracts/feirari/FuseGuardian.sol (hunk 2 above)
Comment (ditchfieldcaleb) [reply contracts/feirari/FuseGuardian.sol:47]: @transmissions11 does this save gas or is it just a visual / code style thing

Comment (transmissions11) [reply contracts/feirari/FuseGuardian.sol:47]: honestly cant remember, this is why y'all need gas snapshotting infra :)

File: proposals/description/fip_60.ts [hunk 6]
```diff
@@ -44,6 +44,51 @@ const fip_60: ProposalDescription = {
...
+      arguments: ['{fuseGuardian}'],
+      description: 'Set Fuse borrow cap guardian'
+    },
+    {
+      target: 'fuseGuardian',
+      values: '0',
+      method: '_setMarketSupplyCapsByUnderlying(address[],uint256[])',
+      arguments: [
+        ['{curveD3pool}', '{curve3Metapool}', '{gUniFeiDaiLP}'],
+        ['25000000000000000000000000', '25000000000000000000000000', '2500000000000000000000000000']
```
Comment (ditchfieldcaleb) [inline proposals/description/fip_60.ts:88]: We should come up with a better way to spell out amounts than an inline string, it's bound to cause a mistake eventually

File: contracts/feirari/FuseGuardian.sol [hunk 7]
```diff
@@ -0,0 +1,109 @@
...
+        require(cToken != address(0), "cToken doesn't exist");
+        _setBorrowPaused(CToken(cToken), state);
+    }
+
+    function _setBorrowPaused(CToken cToken, bool state) public isGovernorOrGuardianOrAdmin returns (bool) {
+        return comptroller._setBorrowPaused(cToken, state);
+    }
+
+    function _setTransferPaused(bool state) public isGovernorOrGuardianOrAdmin returns (bool) {
+        return comptroller._setSeizePaused(state);
```
Comment (ElliotFriedman) [inline contracts/feirari/FuseGuardian.sol:103]: should call `_setTransferPaused ` not `_setSeizePaused`

File: contracts/feirari/FuseGuardian.sol [hunk 8]
```diff
@@ -0,0 +1,109 @@
...
+    /**
+     * @notice Admin function to change the Pause Guardian
+     * @param newPauseGuardian The address of the new Pause Guardian
+     * @return uint 0=success, otherwise a failure. (See enum Error for details)
+     */
+    function _setPauseGuardian(address newPauseGuardian) public onlyGovernor returns (uint) {
+        return comptroller._setPauseGuardian(newPauseGuardian);
+    }
+
+    function _setMintPausedByUnderlying(address underlying, bool state) external {
```
Comment (ElliotFriedman) [inline contracts/feirari/FuseGuardian.sol:82]: I don't like how this function calls a public function that has a modifier on it and has no modifier on itself in the function signature. It obfuscates who is allowed to call it.

Prefer splitting out the _setMintPaused into an internal helper function and permissioning both functions.

**Review by ElliotFriedman — DISMISSED** [review]

File: contracts/feirari/FuseGuardian.sol (hunk 1 above)
Comment (Joeysantoro) [reply contracts/feirari/FuseGuardian.sol:42]: I'm probably going to leave as-is given this is an admin function and I'd rather optimize for readability than gas

File: contracts/feirari/FuseGuardian.sol (hunk 2 above)
Comment (Joeysantoro) [reply contracts/feirari/FuseGuardian.sol:47]: same point

File: contracts/feirari/FuseGuardian.sol (hunk 8 above)
Comment (Joeysantoro) [reply contracts/feirari/FuseGuardian.sol:82]: Great callout, fixed

File: proposals/description/fip_60.ts (hunk 6 above)
Comment (Joeysantoro) [reply proposals/description/fip_60.ts]: can probably just do ethers weiperether

Comment (Joeysantoro) [reply proposals/description/fip_60.ts]: I do think we're covered by invariant checking in the validate step

File: contracts/feirari/FuseGuardian.sol [hunk 9]
```diff
@@ -0,0 +1,117 @@
...
+      * @notice Set the given supply caps for the given cToken markets. Supplying that brings total underlying supply to or above supply cap will revert.
+      * @dev Admin or borrowCapGuardian function to set the supply caps. A supply cap of 0 corresponds to unlimited supplying.
+      * @param cTokens The addresses of the markets (tokens) to change the supply caps for
+      * @param newSupplyCaps The new supply cap values in underlying to be set. A value of 0 corresponds to unlimited supplying.
+      */
+    function _setMarketSupplyCaps(CToken[] memory cTokens, uint[] calldata newSupplyCaps) public isGovernorOrGuardianOrAdmin {
+        comptroller._setMarketSupplyCaps(cTokens, newSupplyCaps);
+    }
+
+    function _setMarketSupplyCapsByUnderlying(address[] calldata underlyings, uint[] calldata newSupplyCaps) external {
```
Comment (ElliotFriedman) [inline contracts/feirari/FuseGuardian.sol:36]: I don't like how this function calls a public function that has a modifier on it and has no modifier on itself in the function signature. It obfuscates who is allowed to call it.

File: contracts/feirari/FuseGuardian.sol (hunk 2 above)
Comment (ElliotFriedman) [reply contracts/feirari/FuseGuardian.sol:47]: I've seen gas savings from using a named parameter in the return value

**Pull request merged**